lib/controllers/hosted-skill-controller/helper.js

```diff
@@ -81,3 +95,74 @@ function handleSkillStatus(response, skillId, callback) {
         callback(`${errMessage}${errReason}Infrastructure provision for the hosted skill failed. Please try again.`);
     }
 }
+
+function checkScript(scriptUrl, filePath, callback) {
+    if (!fs.existsSync(filePath)) {
+        return this.downloadScriptFromS3(scriptUrl, filePath, (err) => callback(err || null));
+    }
+    isScriptUpdated(scriptUrl, filePath, (err, isUpdated) => {
+        if (err) {
+            return callback(err);
+        }
+        if (!isUpdated) {
+            return callback(null);
+        }
+        this.downloadScriptFromS3(scriptUrl, filePath, (downloadErr) => callback(downloadErr || null));
+    });
+}
+
+function isScriptUpdated(scriptUrl, filePath, callback) {
```

Comment:
When this going to run? Do we need all this extra logic with etags?

Reply:
It is going to run when checking scripts, and the `checkScript` method is called whenever `ask new` an AHS, `ask init` an AHS and `ask util upgrade-project` an AHS project.
I think the `isScriptUpdated` is a better design for two reasons. Firstly, It's getting frequently we check the scripts, but in over 99% cases, we don't have to download the scripts, thus, I think we can check the resource's metadata before retrieving the entire resource.   Secondly, we are using `Head` http request to get metadata to check the etag, it's a cheaper request. We may not see any faster between `HEAD` and `GET` in this case since the scripts are small, but I believe that to check script if it is updated before downloading the scripts is a more reasonable design.

Reply:
sorry. it still feels overkill for 3-4 kb files.